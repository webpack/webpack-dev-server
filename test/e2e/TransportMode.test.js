'use strict';

const testServer = require('../helpers/test-server');
const config = require('../fixtures/client-config/webpack.config');
const runBrowser = require('../helpers/run-browser');
const port = require('../ports-map').TransportMode;
const {
  initConsoleDelay,
  awaitServerCloseDelay,
} = require('../helpers/puppeteer-constants');

describe('transportMode client', () => {
  const modes = [
    {
      title: 'sockjs',
      options: {
        hot: false,
        transportMode: 'sockjs',
      },
    },
    {
      title: 'ws',
      options: {
        hot: false,
        transportMode: 'ws',
      },
    },
    {
      title: 'custom client',
      options: {
        hot: false,
        transportMode: {
          server: 'sockjs',
          client: require.resolve(
            '../fixtures/custom-client/CustomSockJSClient'
          ),
        },
      },
    },
  ];

  modes.forEach((mode) => {
    describe(mode.title, () => {
      beforeAll((done) => {
        const options = Object.assign(
          {},
          {
            port,
            host: '0.0.0.0',
          },
          mode.options
        );
        testServer.startAwaitingCompilation(config, options, done);
      });

      describe('on browser client', () => {
        it('logs correctly', (done) => {
          runBrowser().then(({ page, browser }) => {
            const res = [];
            page.goto(`http://localhost:${port}/main`);
            page.on('console', ({ _text }) => {
              res.push(_text);
            });

            page.waitForTimeout(initConsoleDelay).then(() => {
              testServer.close(() => {
                // make sure the client gets the close message
                page.waitForTimeout(awaitServerCloseDelay).then(() => {
                  browser.close().then(() => {
                    for (let i = res.length - 1; i >= 0; i--) {
                      if (res[i] === '[webpack-dev-server] Disconnected!') {
                        break;
                      } else if (
                        res[i] === 'close' ||
                        res[i].includes('net::ERR_CONNECTION_REFUSED') ||
                        // this indicates a WebSocket Error object that was logged
                        res[i].includes('JSHandle@object')
                      ) {
                        // remove additional logging for the now failing connection,
                        // since this could be a variable number of error messages
                        res.splice(i, 1);
                      }
                    }
                    expect(res).toMatchSnapshot();
                    done();
                  });
                });
              });
            });
          });
        });
      });
    });
  });
});
