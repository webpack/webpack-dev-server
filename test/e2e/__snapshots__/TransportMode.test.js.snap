// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transportMode client custom client on browser client logs correctly 1`] = `
Array [
  "Hey.",
  "open",
  "liveReload",
  "[webpack-dev-server] Live Reloading enabled.",
  "hash",
  "ok",
  "close",
  "[webpack-dev-server] Disconnected!",
]
`;

exports[`transportMode client sockjs on browser client logs correctly 1`] = `
Array [
  "Hey.",
  "[webpack-dev-server] Live Reloading enabled.",
  "[webpack-dev-server] Disconnected!",
]
`;

exports[`transportMode client ws on browser client logs correctly 1`] = `
Array [
  "Hey.",
  "[webpack-dev-server] Live Reloading enabled.",
  "[webpack-dev-server] Disconnected!",
]
`;
